### PR TITLE
Use `cast_to_bw` and `cast_from_bw` for some type conversions

### PR DESCRIPTION
**Pull request by nlinker**

### CODE REVIEW

File: src/BulletType.hpp [hunk 1]
```diff
@@ -1,14 +1,3 @@
...
-inline BWAPI::BulletType bullettype_to_bw(BulletType type) {
-    return BWAPI::BulletType(type.id);
-}
-
-inline BulletType bullettype_from_bw(BWAPI::BulletType type) {
-    BulletType self;
-    self.id = type.getID();
-    return self;
-}
+// moved to Cast.hpp
```
Comment (kpp) [inline src/BulletType.hpp]: drop the file, don't keep it

File: src/Cast.hpp [hunk 2]
```diff
@@ -6,6 +6,12 @@
 #include <BWAPI/UnitType.h>
 #include <BWAPI/Event.h>
 #include <BWAPI/EventType.h>
+#include <BWAPI/BulletType.h>
```
Comment (kpp) [inline src/Cast.hpp]: `#include <BWAPI.h>` for all possible headers

File: src/Input.hpp [hunk 3]
```diff
@@ -1,24 +1,3 @@
...
-inline BWAPI::MouseButton mousebutton_to_bw(MouseButton button) {
-    return static_cast<BWAPI::MouseButton>(button.id);
-}
-
-inline MouseButton mousebutton_from_bw(BWAPI::MouseButton button) {
-    MouseButton self;
-    self.id = static_cast<int>(button);
-    return self;
-}
+// moved to Cast.hpp
```
Comment (kpp) [inline src/Input.hpp]: drop the file, don't keep it

File: src/Order.hpp [hunk 4]
```diff
@@ -1,14 +1,3 @@
...
-inline BWAPI::Order order_to_bw(Order order) {
-    return BWAPI::Order(order.id);
-}
-
-inline Order order_from_bw(BWAPI::Order order) {
-    Order self;
-    self.id = order.getID();
-    return self;
-}
+// moved to Cast.hpp
```
Comment (kpp) [inline src/Order.hpp]: drop the file, don't keep it

File: src/Position.hpp [hunk 5]
```diff
@@ -1,37 +1,3 @@
...
-    return BWAPI::WalkPosition(position.x, position.y);
-}
-
-inline WalkPosition walkposition_from_bw(BWAPI::WalkPosition position) {
-    WalkPosition self;
-    self.x = position.x;
-    self.y = position.y;
-    return self;
-}
+// moved to Cast.hpp
```
Comment (kpp) [inline src/Position.hpp]: drop the file, don't keep it

File: src/Region.cpp [hunk 6]
```diff
@@ -3,7 +3,6 @@
 #include <BWAPI/Regionset.h>
```
Comment (kpp) [inline src/Region.cpp]: Don't keep unnecessary includes, as they will be included from IteratorImpl -> Cast -> BWAPI.h

File: src/UnitCommand.hpp [hunk 7]
```diff
@@ -1,30 +1,3 @@
...
-    UnitCommand self;
-    self.unit = reinterpret_cast<Unit*>(command.unit);
-    self.type = unitcommandtype_from_bw(command.type);
-    self.target = reinterpret_cast<Unit*>(command.target);
-    self.x = command.x;
-    self.y = command.y;
-    self.extra = command.extra;
-    return self;
-}
+// moved to Cast.hpp
```
Comment (kpp) [inline src/UnitCommand.hpp]: drop the file, don't keep it

File: src/WeaponType.hpp [hunk 8]
```diff
@@ -1,14 +1,3 @@
...
-inline BWAPI::WeaponType weapontype_to_bw(WeaponType type) {
-    return BWAPI::WeaponType(type.id);
-}
-
-inline WeaponType weapontype_from_bw(BWAPI::WeaponType type) {
-    WeaponType self;
-    self.id = type.getID();
-    return self;
-}
+// moved to Cast.hpp
```
Comment (kpp) [inline src/WeaponType.hpp]: drop the file, don't keep it

File: src/Region.cpp (hunk 6 above)
Comment (nlinker) [inline src/Region.cpp]: Isn't this too complicated to keep the DAG in the head?

File: src/AIModule.cpp [hunk 9]
```diff
@@ -1,7 +1,6 @@
 #include <AIModule.h>
 #include <BWAPI/AIModule.h>
```
Comment (kpp) [inline src/AIModule.cpp]: Delete all BWAPI/ headers as they are in Cast.hpp

File: src/Bullet.cpp [hunk 10]
```diff
@@ -1,8 +1,6 @@
 #include <Bullet.h>
 #include <BWAPI/Bullet.h>
```
Comment (kpp) [inline src/Bullet.cpp]: del

File: src/Cast.hpp [hunk 11]
```diff
@@ -116,12 +201,100 @@ struct CastRev<UnitType> {
...
+//----------------------------------------------------
+
+template<>
 struct Cast<BWAPI::EventType::Enum, EventType> {
     typedef BWAPI::EventType::Enum BwType;
     typedef EventType              TxType;
 
     inline static TxType from_bw(BwType bw) {
-        return EventType { bw };
+        return EventType {bw};
```
Comment (kpp) [inline src/Cast.hpp]: better use explicit `::getID()` instead of implicit cast to int

File: src/Cast.hpp [hunk 12]
```diff
@@ -168,6 +385,41 @@ template<>
...
+        return self;
+    }
+    inline static BwType to_bw(TxType tx) {
+        return BWAPI::UnitCommand(
+            cast_to_bw(tx.unit),
+            cast_to_bw(tx.type),
+            cast_to_bw(tx.target),
+            tx.x,
+            tx.y,
+            tx.extra);
```
Comment (kpp) [inline src/Cast.hpp]: `);` should be on the next line

**Review by kpp — APPROVED** [review]

**Review by 0x7CFE — APPROVED** [review]
LGTM

**Pull request merged**